cidc_schemas/json_validation.py

```diff
@@ -159,24 +196,21 @@ def _ensure_in_doc_ref(
             # if we're in a "simple" value context,
             # we can't possibly match ref_path_pattern
             return False
+        
+        # If there are no cached values for this ref path pattern, collect them
+        if ref_path_pattern not in self.in_doc_refs:
+            vals = self._get_values_for_path_pattern(ref_path_pattern, doc)
+            if len(vals) == 0:
+                # There are no values matching this pattern, so there's
+                # no way `ref` can match a value with this pattern.
+                return False
```

Comment:
I think you skip saving to the cache here, which will lead to recalculating `_get_values_for_path_pattern` every time, 
```suggestion
                self.in_doc_refs[ref_path_pattern] = vals
                return False
```

Reply:
As discussed, because of the recursive nature of `iter_errors`, this actually doesn't work.